packages/server/src/fhir/binary.ts

```diff
@@ -41,70 +44,44 @@ async function handleBinaryWriteRequest(req: Request, res: Response): Promise<vo
-  // """
-  let binaryContentSpecialCase = false;
-
   if (contentType === ContentType.FHIR_JSON) {
     const str = await readStreamToString(stream);
+    binarySource = str;
+
     try {
       // The binary handler does *not* use Express body-parser in order to support raw binary data.
       // Therefore, we need to manually parse the body stream as JSON.
       const body = JSON.parse(str);
-      if (isResource(body) && body.resourceType === 'Binary' && body.id === id) {
+      if (isResource(body) && body.resourceType === 'Binary' && (!id || body.id === id)) {
         // Special case where the content is actually a Binary resource.
-        binary = body as Binary;
-        binaryContentSpecialCase = true;
-      } else {
-        // We have already consumed the stream, so we need to create a new one.
-        // Instead, use the original string as the source.
-        binarySource = str;
+        // From the spec: https://hl7.org/fhir/R4/binary.html#rest
+        //
+        // """
+        //   When binary data is written to the server (create/update - POST or PUT),
+        //   the data is accepted as is and treated as the content of a Binary,
+        //   including when the content type is "application/fhir+xml" or "application/fhir+json",
+        //   except for the special case where the content is actually a Binary resource.
+        // """
+        const resource = body as Binary;
+        const binary = await (create ? repo.createResource<Binary>(resource) : repo.updateResource<Binary>(resource));
+        await sendResponse(req, res, create ? created : allOk, binary);
+        return;
       }
     } catch (err) {
       // If the JSON is invalid, then it is not eligible for the special case.
       getLogger().debug('Invalid JSON', { error: err });
-      binarySource = str;
     }
   }
 
-  if (!binary) {
-    const securityContext = req.get('X-Security-Context');
-    binary = {
-      resourceType: 'Binary',
-      id,
-      contentType,
-      securityContext: securityContext ? { reference: securityContext } : undefined,
-    };
-  }
-
-  let outcome: OperationOutcome;
-
-  if (create) {
-    binary = await ctx.repo.createResource<Binary>(binary);
-    outcome = created;
-  } else {
-    binary = await ctx.repo.updateResource<Binary>(binary);
-    outcome = allOk;
-  }
-
-  if (!binaryContentSpecialCase) {
-    const filename = req.query['_filename'] as string | undefined;
-    await getBinaryStorage().writeBinary(binary, filename, contentType, binarySource);
-  }
-
-  await sendResponse(req, res, outcome, {
-    ...binary,
-    url: getBinaryStorage().getPresignedUrl(binary),
+  const binary = await uploadBinaryData(repo, binarySource, {
+    id,
+    contentType,
+    filename: req.query['_filename'] as string | undefined,
+    securityContext: req.get('X-Security-Context'),
   });
+
+  await sendResponse(req, res, create ? created : allOk, binary);
```

Comment:
I'm having a hard time following all of the refactoring.  It seems like the main objective was to refactor out a `uploadBinaryData` helper?